[lldb][AArch64] Fix arm64 hardware breakpoint/watchpoint to arm32 process.

b10902118 opened the pull request:
When debugging arm32 process on arm64 machine, arm64 lldb-server will use `NativeRegisterContextLinux_arm`, but the server should keep using 64-bit ptrace commands for hardware watchpoint/breakpoint, even when debugging a 32-bit tracee.

See: https://github.com/torvalds/linux/commit/5d220ff9420f8b1689805ba2d938bedf9e0860a4

There have been many conditional compilation handling arm32 tracee on arm64, but this one is missed out.

To reuse the 64-bit implementation, I separate the shared code from `NativeRegisterContextLinux_arm64.cpp` to `NativeRegisterContextLinux_arm64dbreg.cpp`, with other adjustments to share data structures of debug registers.

labath commented:
Your analysis is correct -- the kernel uses the bitwidth of the tracer (not the tracee) when deciding whether to use the compatibility interface. However, copying the code sounds like a fairly fragile solution. It'd be better to figure out a way to share the code somehow.

The x86 classes don't have this problem because NativeRegisterContextLinux_x86_64 knows how to debug both 32 and 64-bit processes. That may not be the right solution here, but it's one thing you can use for inspiration.

b10902118 commented:
> > the test
> 
> By this you mean it skips any test that requires hardware breakpoints? Or you mean that in the general course of using lldb it doesn't use hardware breakpoints because of this?
> 
> (I assume you could make the test suite launch all debugees as AArch32 but seems unlikely you went to that effort)

Oh sorry please ignore that. I mixed lldb's test up in other sources (probably android's), making a wrong comment.

b10902118 commented:
> > So the solution is copying the implementation in NativeRegisterContextLinux_arm64.cpp.
> 
> Yes, try putting the 64-bit code in a separate file in Process/Linux then using it in both places.
> 
> Currently arm64 includes the arm header, for reasons I am not sure, so if there is a dependency, it's going the wrong way for us to use it for this.

The include of arm header is just for the factory function `CreateHostNativeRegisterContextLinux` to return `NativeRegisterContextLinux_arm` for arm32 process, tested by removing only that part and compiling.

DavidSpickett commented:
> I am wondering whether the arm-on-arm64 case isn't actually more similar to arm64-on-arm64 than it is to arm32-onarm32.

If arm64 did not have umpteen kinds of registers maybe, but going by the header the only common bit would be the detection and writing of breakpoint registers (for some reason I thought we would also read them back but no of course we don't need to do that).

I haven't tried actually debugging something though so maybe there's more here that doesn't work.

...if there is a way to hack the test suite config to produce AArch32 that's one way to find that out. CMake options LLDB_TEST_COMPILER and LLDB_TEST_ARCH maybe. See how terrible the results are.

b10902118 commented:
I have run API tests on watchpoint (`-p '.*Watch.*'`, not sure how to do it correctly, running all takes too long), and it maintains the same 7 failure so I assmue my modification won't break things.

I also have updated the commit message. To merge, please use the new message. (I have no commit access)

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinuxArm64Shared.cpp:
```diff
@@ -0,0 +1,63 @@
+#include "NativeRegisterContextLinuxArm64Shared.h"
+
+using namespace lldb_private::process_linux::arm64;
+
+namespace lldb_private {
+namespace process_linux {
+namespace arm64 {
```
You don't need `using namespace` if you're then going to open the same namespaces.

Here I would just write out the fully qualified names and have `using namespace` if you still need it after doing that. `lldb_private::proccess_linux::arm64::ReadHardwareDebugInfo` for example.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinuxArm64Shared.cpp:
```diff
@@ -0,0 +1,63 @@
+namespace arm64 {
+
+namespace {
+Status ReadHardwareDebugInfoHelper(int regset, ::pid_t tid,
```
llvm style is to use a static function rather than anonymous namespaces. https://llvm.org/docs/CodingStandards.html#restrict-visibility

Though it doesn't actually show an example of doing it, I will fix that. It does say right at the end to use static.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinuxArm64Shared.h:
```diff
@@ -0,0 +1,23 @@
+
+} // namespace arm64
+} // namespace process_linux
+} // namespace lldb_private
```
Add a newline at the end of the file.

I'm of the camp that wonders why this is needed in this day and age, but it does trip up some tools and it's easy to fix.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinuxArm64Shared.cpp:
```diff
@@ -0,0 +1,63 @@
+#include "NativeRegisterContextLinuxArm64Shared.h"
```
Also the top of this file needs a comment explaining why it needed to be split out.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm.h:
```diff
@@ -74,8 +75,10 @@ class NativeRegisterContextLinux_arm : public NativeRegisterContextLinux {
 
-  // Debug register type select
-  enum DREGType { eDREGTypeWATCH = 0, eDREGTypeBREAK };
+  using DREGType = NativeRegisterContextDBReg::DREGType;
+  static const DREGType eDREGTypeBREAK = DREGType::eDREGTypeBREAK;
+  static const DREGType eDREGTypeWATCH = DREGType::eDREGTypeWATCH;
+  using DREG = NativeRegisterContextDBReg::DREG;
```
I don't see the need for `static const DREGType eDREGTypeBREAK` and I think it would be clearer to fully qualify DREGType and DREG rather than `using` them.

b10902118 replied:
I did that just for modifying less files, not sure whether this should be a concern.

I am considering porting arm to use NativeRegisterContextDBReg, after this it will be not fully qualified again.

DavidSpickett replied:
> I am considering porting arm to use NativeRegisterContextDBReg, after this it will be not fully qualified again.

If it's the class that LoongArch and AArch64 use, I might have a tree where I did this. I never found the time to test it with Arm/Thumb options though.

Please fully qualify the names for now and I'll look for those changes of mine.

DavidSpickett replied:
I do have those changes still.

I suggest that we go ahead with your change here, as I don't want to derail you from the problem you came here to fix.

Once that's in, I will rebase my changes on top. I needed to read them through carefully anyway, so I can do both at the same time. I want to be sure I don't drop any Arm/Thumb handling details along the way.

b10902118 replied:
Got it. I know thumb breakpoint address requires an additional 1 offset, at least in gdb, kind of annoying.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm.h:
```diff
@@ -102,18 +105,8 @@ class NativeRegisterContextLinux_arm : public NativeRegisterContextLinux {
-                           // occurred.
-    lldb::addr_t real_addr; // Address value that should cause target to stop.
-    uint32_t control;       // Breakpoint/watchpoint control value.
-    uint32_t refcount;      // Serves as enable/disable and reference counter.
```
Switching to the other definition removes `refcount` which appears to be unused anyway.

Send a separate PR to remove that field from all definitions so the bots can test that assertion in isolation.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.h:
```diff
@@ -0,0 +1,29 @@
+//===----------------------------------------------------------------------===//
+
+// These functions are split out to be reused in NativeRegisterContextLinux_arm,
+// for supporting debugging 32bit processes on arm64.
```
Add something like "When debugging 32-bit process on Arm64, 64-bit ptrace interfaces must be used.". To me, this is the key detail.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm.cpp:
```diff
@@ -737,12 +744,17 @@ Status NativeRegisterContextLinux_arm::ReadHardwareDebugInfo() {
+#else  // __aarch64__
+  return arm64::ReadHardwareDebugInfo(m_thread.GetID(), m_max_hwp_supported,
+                                      m_max_hbp_supported);
+#endif // __arm__
```
These final #endif we tend to write the `if` as well like `ifdef __arm__`.

DavidSpickett replied:
If it's something very complex then yeah we simplify it because you should just look for the original `if` and read that.

DavidSpickett replied:
It's not automatically enforced so sometimes things are inconsistent yes. This is new code though so it's fine to be more precise.

DavidSpickett commented on lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp:
```diff
@@ -0,0 +1,71 @@
+Status lldb_private::process_linux::arm64::WriteHardwareDebugRegs(
+    int hwbType, ::pid_t tid, uint32_t max_supported,
+    const std::array<NativeRegisterContextDBReg::DREG, 16> &regs) {
+  struct iovec ioVec;
```
Move your variables to be next to their first use, and add some blank lines between the "steps" of this process. For example one after memset and one after you calculate iov_len.

DavidSpickett reviewed (commented):
Some small comments and you'll need to rebase when the cleanup PR lands. It's good to go in once that's done.

DavidSpickett commented:
From the PR description:
> After tracing linux's compat_ptrace in arch/arm64/kernel/ptrace.c, found that arm64 lldb-server should just keep using the ptrace commands for 64bit tracees.

Reword this so it's a bit clearer. I get what you mean now, but I had to read it several times.

"lldb-server should keep using the ptrace commands intended for 64-bit bit tracees, even when debugging a 32-bit tracee"

DavidSpickett commented:
Linux CI failure is:
```
2025-07-28T08:45:11.0284286Z sccache /opt/llvm/bin/clang++ -DEXPERIMENTAL_KEY_INSTRUCTIONS -DGTEST_HAS_RTTI=0 -DHAVE_ROUND -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -D__SHORT_FILE__=\"NativeRegisterContextLinux_arm64dbreg.cpp\" -I/home/gha/actions-runner/_work/llvm-project/llvm-project/build/tools/lldb/source/Plugins/Process/Linux -I/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux -I/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/include -I/home/gha/actions-runner/_work/llvm-project/llvm-project/build/tools/lldb/include -I/home/gha/actions-runner/_work/llvm-project/llvm-project/build/include -I/home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/include -I/usr/include/python3.12 -I/home/gha/actions-runner/_work/llvm-project/llvm-project/llvm/../clang/include -I/home/gha/actions-runner/_work/llvm-project/llvm-project/build/tools/lldb/../clang/include -I/home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source -I/home/gha/actions-runner/_work/llvm-project/llvm-project/build/tools/lldb/source -gmlt -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -Wno-unknown-pragmas -Wno-strict-aliasing -Wno-vla-extension -O3 -DNDEBUG -std=c++17  -fno-exceptions -funwind-tables -fno-rtti -UNDEBUG -MD -MT tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_arm64dbreg.cpp.o -MF tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_arm64dbreg.cpp.o.d -o tools/lldb/source/Plugins/Process/Linux/CMakeFiles/lldbPluginProcessLinux.dir/NativeRegisterContextLinux_arm64dbreg.cpp.o -c /home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp
2025-07-28T08:45:11.0299743Z /home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp:26:29: error: variable has incomplete type 'struct user_hwdebug_state'
2025-07-28T08:45:11.0301314Z    26 |   struct user_hwdebug_state dreg_state;
2025-07-28T08:45:11.0301735Z       |                             ^
2025-07-28T08:45:11.0302987Z /home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp:26:10: note: forward declaration of 'user_hwdebug_state'
2025-07-28T08:45:11.0304281Z    26 |   struct user_hwdebug_state dreg_state;
2025-07-28T08:45:11.0304681Z       |          ^
2025-07-28T08:45:11.0305917Z /home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp:56:29: error: variable has incomplete type 'struct user_hwdebug_state'
2025-07-28T08:45:11.0307300Z    56 |   struct user_hwdebug_state dreg_state;
2025-07-28T08:45:11.0307782Z       |                             ^
2025-07-28T08:45:11.0309023Z /home/gha/actions-runner/_work/llvm-project/llvm-project/lldb/source/Plugins/Process/Linux/NativeRegisterContextLinux_arm64dbreg.cpp:56:10: note: forward declaration of 'user_hwdebug_state'
2025-07-28T08:45:11.0310341Z    56 |   struct user_hwdebug_state dreg_state;
2025-07-28T08:45:11.0310890Z       |          ^
2025-07-28T08:45:11.0311164Z 2 errors generated.
```

b10902118 commented:
I am fixing the ci failure. Just wonder how you guys prevent this? I can build locally and clangd also didn't complain in my editor.

DavidSpickett commented:
Yeah that is odd. Is it because `NativeRegisterContextLinux_arm64dbreg.cpp` is being built even when we're not compiling lldb-server for AArch64 or Arm? Perhaps because you are on an AArch64 machine that's why it works for you?

b10902118 committed:
Add arm64 architecture guard. Should fix ci build.

b10902118 commented:
> Edit: because the CI is x86_64.

Thanks for pointing that out. The failure is probably cause by the missing `#if defined(__arm64__) || defined(__aarch64__)`. I finally understand why this is needed.

b10902118 committed:
Stress on arm64 servers have to debug 32-bit processes with 64-bit ptrace.

b10902118 committed:
Detail #endif comment.

b10902118 committed:
Add spaces between WriteHardwareDebugRegs steps.

b10902118 committed:
Fully qualify DREG and DREGType.

b10902118 committed:
[lldb][AArch64] Fix arm64 hardware breakpoint/watchpoint to arm32 process.

This bug skips the test because the wrong ptrace call to detect avaliable hardware/breakpoint number that resuls in 0.

After tracing linux's compat_ptrace in arch/arm64/kernel/ptrace.c, found that arm64 lldb-server should just keep using the ptrace commands for 64bit tracees. See:
https://github.com/torvalds/linux/commit/5d220ff9420f8b1689805ba2d938bedf9e0860a4

So the solution is copying the implementation in NativeRegisterContextLinux_arm64.cpp.

b10902118 committed:
[lldb][AArch64] Fix hardware breakpoint/watchpoint to arm32 debugee on arm64.

When debugging arm32 process on arm64 machine, arm64 lldb-server will use `NativeRegisterContextLinux_arm`, but it should use 64-bit ptrace commands for hardware watchpoint/breakpoint.
See: https://github.com/torvalds/linux/commit/5d220ff9420f8b1689805ba2d938bedf9e0860a4

There have been many conditional compilation handling arm32 debuggees on arm64, but this one is missed out.

To reuse the 64-bit implementation, I separate the shared code from `NativeRegisterContextLinux_arm64.cpp` to `NativeRegisterContextLinuxArm64Shared.cpp`, with other adjustments to share data structures of debug registers.

b10902118 committed:
Rename files to dbreg.

b10902118 committed:
Fix style, license headers, and includes.

move includes to .cpp and remove <sys/ptrace> that causes error when build.

DavidSpickett commented:
Ignore that failure, it "fixed" itself.

DavidSpickett commented:
> Once that's in, I will rebase my changes on top. I needed to read them through carefully anyway, so I can do both at the same time. I want to be sure I don't drop any Arm/Thumb handling details along the way.

I have rebased those changes to include this and found https://github.com/llvm/llvm-project/pull/151973 along the way. Once that's in I'll put up a PR to port Arm to NativeRegisterContextDBReg.

DavidSpickett commented:
https://github.com/llvm/llvm-project/pull/152284
